### PR TITLE
Add support for monitoring aiohttp based applications

### PR DESCRIPTION
**Pull request by michaelyaakoby**
@amenezes extended Pyctuator to support monitoring of aiohttp based applications by Spring Boot Admin.
Thanks

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -139,6 +139,36 @@ The application will automatically register with Spring Boot Admin upon start up
...
+The following example is complete and should run as is.
+
+```python
+from aiohttp import web
+from pyctuator.pyctuator import Pyctuator
+
+app = web.Application()
+routes = web.RouteTableDef()
+
+@route.get("/")
```
Comment (MatanRubin) [inline README.md:152]: should this be `@routes.get("/")`?

File: examples/aiohttp/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# aiohttp example
+This example demonstrates the integration with the [aiohttp](https://flask.palletsprojects.com/).
```
Comment (MatanRubin) [inline examples/aiohttp/README.md:2]: wrong link for aiohttp

File: examples/aiohttp/README.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+## Running the example
+1. Start an instance of SBA (Spring Boot Admin):
+    ```sh
+    docker run -p 8082:8082 --rm michayaak/spring-boot-admin:2.2.2
+    ```
+2. Once Spring Boot Admin is running, you can run the examples as follow:
+    ```sh
+    cd examples/aiohttp
+    poetry install
+    poetry run python -m aiohttp_example_app
```
Comment (MatanRubin) [inline examples/aiohttp/README.md:13]: I tried running this example following the instructions, and got the following error:

```
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/matan.r/Code/SolarEdge/pyctuator/examples/aiohttp/aiohttp_example_app.py", line 31, in <module>
    app_description="Demonstrate Spring Boot Admin Integration with aiohttp",
  File "/Users/matan.r/Code/SolarEdge/pyctuator/examples/aiohttp/.venv/lib/python3.7/site-packages/pyctuator/pyctuator.py", line 122, in __init__
    raise EnvironmentError("No framework was found that is matching the target app"
OSError: No framework was found that is matching the target app(is it properly installed and imported?)
```

File: pyctuator/impl/aiohttp_pyctuator.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+                web.get("/pyctuator/dump", get_thread_dump),
+                web.get("/pyctuator/threaddump", get_thread_dump),
+                web.get("/pyctuator/logfile", get_logfile),
+                web.get("/pyctuator/trace", get_httptrace),
+                web.get("/pyctuator/httptrace", get_httptrace),
+            ]
+        )
+        app.middlewares.append(intercept_requests_and_responses)
+
+    def _to_dict(self, data: object) -> Optional[dict]:
```
Comment (MatanRubin) [inline pyctuator/impl/aiohttp_pyctuator.py:158]: So if the object is already a `dict`, this just swallows the data? Seems like odd behavior. Maybe raise an exception? If so, the implementation would just be `dataclasses.asdict(data)`, becuase I think `asdict` checks that the data is a dataclass.

File: pyctuator/impl/aiohttp_pyctuator.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        if dataclasses.is_dataclass(data):
+            return dataclasses.asdict(data)
+        return None
+
+    def _datetime_serializer(self, value: Any) -> Optional[str]:
```
Comment (MatanRubin) [inline pyctuator/impl/aiohttp_pyctuator.py:163]: If this only handles datetimes, why have it accept `Any`?

File: pyctuator/pyctuator.py [hunk 6]
```diff
@@ -51,6 +51,8 @@ def __init__(
 
         * Flask - `app` is an instance of `flask.app.Flask`
 
+        * aiohttp - `app`is an instance of `aiohttp.web.Application`
```
Comment (MatanRubin) [inline pyctuator/pyctuator.py:54]:
```suggestion
        * aiohttp - `app` is an instance of `aiohttp.web.Application`
```

File: examples/aiohttp/README.md (hunk 3 above)
Comment (MatanRubin) [reply examples/aiohttp/README.md:13]: I looked at the code and it seems like it should work. Does it work on your setup?

File: README.md (hunk 1 above)
Comment (michaelyaakoby) [reply README.md:152]: True (and IntelliJ showed a warning... cool)

File: examples/aiohttp/README.md (hunk 3 above)
Comment (michaelyaakoby) [reply examples/aiohttp/README.md:13]: Yes, its working for me.
Dod you:
```
poetry install --extras aiohttp
```

File: examples/aiohttp/README.md (hunk 2 above)
Comment (michaelyaakoby) [reply examples/aiohttp/README.md:2]: True, fixed

File: pyctuator/impl/aiohttp_pyctuator.py (hunk 4 above)
Comment (michaelyaakoby) [reply pyctuator/impl/aiohttp_pyctuator.py:158]: The issue of jsonifying a dataclass is more involved as `dataclasses.asdict` doesn't (and shouldn't) cascade into child dataclass members.

I don't think throwing an exception makes it much better, I suggest that we'll leave this "as is" and make sure we have sufficient tests. 
Once a test fails, we can write our own, recursive jsonifier.

File: pyctuator/impl/aiohttp_pyctuator.py (hunk 5 above)
Comment (michaelyaakoby) [reply pyctuator/impl/aiohttp_pyctuator.py:163]: Replaced with custom jsonifying

File: pyctuator/impl/aiohttp_pyctuator.py (hunk 4 above)
Comment (michaelyaakoby) [reply pyctuator/impl/aiohttp_pyctuator.py:158]: replaced with custom jsonifying

**Comment by michaelyaakoby** [comment]
closes #17

File: examples/aiohttp/README.md (hunk 3 above)
Comment (MatanRubin) [reply examples/aiohttp/README.md:13]: I did not, and why should I? This is an aiohttp example, it doesn't make sense for aiohttp to be an optional dependency

Comment (michaelyaakoby) [reply examples/aiohttp/README.md:13]: So the issue is that the example depends on a version of pyctuator that wasn't yet released...

To workaround this, you can use https://python-poetry.org/docs/dependency-specification/#path-dependencies, so you change the dependency in `pyproject.toml` to:
```toml
pyctuator = { path = "../../" }
```

**Comment by amenezes** [comment]
@michaelyaakoby sorry for my absent in these days. I was really busy.

I saw that all tests are ok in the CI are there there any other adjustments needed?

**Comment by michaelyaakoby** [comment]
@amenezes its alright - we are happy you've contributed aiohttp support.

**Pull request merged**